[PATCH v10] api: ipsec: test sa update

asasidharan opened the pull request:
With IPSEC implementations, the protocol specific fields
like sequence number is not exposed to the application.
But for debugging as well as for validation, finer control
of such fields would be required. The proposed API,
odp_ipsec_test_sa_update(), would allow access to such
fields which are not expected to be exposed in the data
path.

psavol commented on include/odp/api/spec/ipsec.h:
```diff
@@ -1601,6 +1635,22 @@ int odp_ipsec_result(odp_ipsec_packet_result_t *result, odp_packet_t packet);
+ * @return 0      On success
+ * @retval <0     On failure
+ */
+int odp_ipsec_debug_sa_update(odp_ipsec_sa_t sa, odp_ipsec_debug_sa_param_t sa_param);
```
Any limitations on usage of this call? When this function can be called: when traffic is flowing, only before starting to use the SA, etc?

Capability needed to tell if implementation support IPsec debug functions.

asasidharan replied:
Updated the following in V2:
- Clarified the usage of the API with the following statement:
```
+ * It is the responsiblity of the application to make sure that no traffic is   
+ * flowing through the SA when the API is called. Otherwise, this operation may 
+ * lead to undefined behavior.
```
- Added additional odp_ipsec_debug_capability_t in the existing odp_ipsec_capability_t structure.

JannePeltonen replied:
I previously suggested having this debug stuff as an optional part of SA creation, but after the addition of the capabilities and the restriction on when the update function may be called, I think I am ok with the separate update function.

psavol replied:
As discussed on the call, may be "test" would be better name to use for this functionality than "debug", as user would use the API for better test coverage (and not always for debugging).

asasidharan replied:
Ack. Updated the PR to use "test" instead of "debug" in v4.

JannePeltonen commented on include/odp/api/spec/ipsec.h:
```diff
@@ -191,6 +233,42 @@ typedef struct odp_ipsec_outbound_config_t {
+			/** Mapping for individual bits */
+			struct {
+				/** Antireplay window top */
+				uint32_t antireplay_window_top  : 1;
```
I am not sure the capabilities should be defined as bit fields as it does not really save memory or cycles. Perhaps these could be just odp_bool_t (or odp_support_t?). Maybe Petri can comment?

psavol replied:
After discussion in call, boolean makes most sense.

asasidharan replied:
Changed antireplay_window_top and seq_num from bit fields to odp_bool_t in v3.

JannePeltonen commented on include/odp/api/spec/ipsec.h:
```diff
@@ -70,6 +70,48 @@ typedef enum odp_ipsec_op_mode_t {
+	/* Antireplay window top */
+	ODP_IPSEC_DEBUG_SA_FIELD_ANTIREPLAY_WINDOW_TOP
+
+} odp_ipsec_debug_sa_param_type_t;
```
The names seem slightly awkward. I think the _FIELD part could be dropped.

> /** Sequence number */
I think this should be "Next sequence number".

> /* Antireplay window top */
This may also be a bit ambiguous. Maybe the comment could be "Highest authenticated sequence number"?

asasidharan replied:
Ack.

JannePeltonen commented on test/validation/api/ipsec/ipsec_test_out.c:
```diff
@@ -1138,6 +1138,66 @@ static void test_out_ipv4_udp_esp_null_sha256(void)
 	ipsec_sa_destroy(sa);
 }
 
+static void test_debug_sa_update_seq_num(void)
```
What is the purpose of this test?

According to the subject line of the commit ("add IPSEC SA sequence number test") this tests something about how SAs handle sequence numbers. Like maybe checking that sequence number starts from one and gets incremented by one for each packet. But it does not test that.

I did not understand the actual text of the commit message.

The name of the test function is ipsec_check_debug_sa_update_seq_num(), leading me to think that the test tests that odp_ipsec_debug_sa_update() somehow works. But it does not test that. odp_ipsec_debug_sa_update() could update the sequence number to some other value than requested or do nothing at all and the test would still pass.

So if the test is not about testing odp_ipsec_debug_sa_update() but about testing ODP IPsec functionality by using  dp_ipsec_debug_sa_update() as part of the test machinery, then the test function should be renamed.

But looking at the code, the only thing the test tests is that IPsec packet is succesfully received if the first received sequence number is 1235. That might test the anti-replay implementation just a little bit, except that the test does not set a non-zero value to antireplay_ws config parameter to ensure that anti-replay is enabled.

So maybe the only thing the test really tests is that the underlying ODP implementation has the SA debug API addition so that compilation does not fail? But I think a simpler test would suffice for that.

asasidharan replied:
> I did not understand the actual text of the commit message.

Ack. Updated the commit message to reflect the actual test being done in v3.


> odp_ipsec_debug_sa_update() could update the sequence number to some other value than requested or do nothing at all and the test would still pass.

Didn't realize this!. My original thought was to have a known vector (NULL cipher with SHA256 auth) which will have the sequence number as well embedded in the result packet. So the memcmp() operation will confirm that the sequence number is updated. But I ended up using AES-GCM here resulting in an incomplete test.

In v3, I have updated the test to check the sequence number present in the IPsec header of the resulting packet. Also I have tried to plug this into the common test framework so that this will be run for all the supported algorithm combinations.

Let me know if this looks good.

JannePeltonen commented on include/odp/api/spec/ipsec.h:
```diff
@@ -71,6 +71,48 @@ typedef enum odp_ipsec_op_mode_t {
+} odp_ipsec_test_sa_param_val_t;
+
+/**
+ * IPSEC TEST SA parameters to allow applications to override unexposed IPSEC
```
The term "override" sounds to me that the value given either permanently overrides the normal state parameter (in which case e.g. sequence number would stay constant) or once (after which the normal state parameter is used again). I think simple "set" would be better here and in other parts of the API description.

asasidharan replied:
Agreed. Updated v5 to use the term "set" instead of "override".

JannePeltonen commented on include/odp/api/spec/ipsec.h:
```diff
@@ -71,6 +71,48 @@ typedef enum odp_ipsec_op_mode_t {
+	/** Highest authenticated sequence number */
+	ODP_IPSEC_TEST_SA_ANTIREPLAY_WINDOW_TOP
+
+} odp_ipsec_test_sa_param_type_t;
```
I am not sure if calling this "type" (instead of "instance" or just parameter) is good. This way we cannot use the same "type" for different parameters, suggesting that it is not really a "type" that we are dealing with.

asasidharan replied:
Agreed. Updated v5 to use "instance" instead of "type".

psavol replied:
Agree, this should indicate selection of what the operation does and what fields from the union (under) is used. Command, operation, select ...

JannePeltonen commented on include/odp/api/spec/ipsec.h:
```diff
@@ -192,6 +234,33 @@ typedef struct odp_ipsec_outbound_config_t {
+ */
+typedef struct odp_ipsec_test_capability_t {
+	/** Support for updating internal parameters of the SA for test */
+	odp_support_t sa_update;
```
This looks redundant. It would seem to me that the individual parameter specific capabilities would be sufficient. If none of them is set, then update is not supported at all.

asasidharan replied:
Agreed. Removed odp_support_t sa_update from capability with v5.

JannePeltonen commented on include/odp/api/spec/ipsec.h:
```diff
@@ -192,6 +234,33 @@ typedef struct odp_ipsec_outbound_config_t {
+		/** Inbound specific internal parameters */
+		struct {
+			/** Highest authenticated sequence number */
+			odp_bool_t antireplay_window_top;
```
I wonder if it is convenient to have this kind of array of booleans instead of e.g. returning an array of supported parameter "type" (or parameter ID) values. Maybe @psavol would like to comment.

If also the actual parameter struct for the update call were defined a bit differently, then the test functionality supported by this API could perhaps be extended by ODP implementation specific parameters without an API change.

asasidharan replied:
>If also the actual parameter struct for the update call were defined a bit differently,

I am not sure if I understood this correctly. Can you describe the layout you have in mind.

JannePeltonen replied:
Let's keep the capability fields as booleans, but let's also simplify the structure by moving all the fields to the top level. The inbound and outbound substructures do not seem to add any value and the field names have to be unique anyway because the parameter to be modified is specified by the same enum that must have unique corresponding values.

asasidharan replied:
Agreed. Made the capability structure to be flat in V6.

JannePeltonen commented on include/odp/api/spec/ipsec.h:
```diff
@@ -71,6 +71,48 @@ typedef enum odp_ipsec_op_mode_t {
+	/** Highest authenticated sequence number */
+	uint64_t antireplay_window_top;
+
+} odp_ipsec_test_sa_param_val_t;
```
I wonder if the API would be more flexible if the above fields were both of the same "generic integer" type. This way most parameters would not need their own fields in this union. OTOH, once someone adds something like and IV or bytes to be used as random data or something like that, we would need an array of bytes. so new parameters may required new values in the union even if we combine the simple types now.

asasidharan replied:
> I wonder if the API would be more flexible if the above fields were both of the same "generic integer" type.

Did you mean something like:
union {
  uint64_t u64;
  uint8_t buffer[MAX_SIZE];
}

JannePeltonen replied:
After discussing with Petri, I think it is better to leave this as proposed, i.e. there is a separate union field for each possible parameter.

asasidharan replied:
Ack.

JannePeltonen commented on include/odp/api/spec/ipsec.h:
```diff
@@ -1715,6 +1789,26 @@ int odp_ipsec_result(odp_ipsec_packet_result_t *result, odp_packet_t packet);
+ *
+ * It is the responsiblity of the application to make sure that no traffic is
+ * flowing through the SA when the API is called. Otherwise, this operation may
+ * lead to undefined behavior.
```
How about using the following description here:

```
* IPsec test API for modifying internal state of an SA.
 *
 * This function is not meant to be used by normal applications but by special
 * test applications that test or debug the operation of the underlying ODP
 * implementation. Calling this function may degrade the performance of the
 * calling thread, other threads or the IPsec implementation in general.
 *
 * Calling this function for an SA at the same time when the SA is used for
 * processing traffic or when the SA is being modified through other parts
 * of IPsec API may result in undefined behaviour.
 *
 * SA state update through this function may not be supported by all ODP
 * implementations, ODP instances or SA instances or at every moment. This
 * function may fail for unspecified reasons even when the capability
 * call indicated support for updating a particular parameter and previous
 * similar calls succeeded.
 ```

asasidharan replied:
Ack.

JannePeltonen commented on test/validation/api/ipsec/ipsec.c:
```diff
@@ -854,6 +869,37 @@ static int ipsec_send_out_one(const ipsec_test_part *part,
+	sa_param.type = ODP_IPSEC_TEST_SA_SEQ_NUM;
+	sa_param.val.seq_num = seq_num;
+
+	CU_ASSERT_EQUAL(0, odp_ipsec_test_sa_update(sa, sa_param));
```
If we allow this function to fail even when the capability call succeeds (so that the capa bit does not promise that update is possible for all SAs at all times), then it is not strictly correct to fail validation test if the call fails. Maybe we could live with the strict check now but it would be better to just notice if the update failed and then turn of the sequence number checking until the update call succeeds again.

asasidharan replied:
Agreed. Update the test to loop on the API call till either it succeeds or one minute timeout happens.

JannePeltonen replied:
I think one minute timeout is excessive as it may repeat for multiple algorithms and then possible slow down CI a lot. Something like one second or less would be better if retrying made sense, but I do not think it does. If the call fails, it probably fails for a reason that does not go away by waiting it out (e.g. for this particular SA seqno cannot be set through this API).

My original point was that since the API says that the function may fail, it is not a violation of the API if the function fails and thus it is not strictly correct for the validation check to fail in that case. Ideally, the validation test would notice that the function failed and then make note of that so that it won't do the subsequent checks that check the outcome of the attempted test operation (that was not done in the first place because the call failed).

asasidharan replied:
Updated the test to skip further check related to seq_num when the API call fails. Also the test would print ">> skipped" for the test configuration which was skipped.

JannePeltonen commented on include/odp/api/spec/ipsec.h:
```diff
@@ -1715,6 +1786,32 @@ int odp_ipsec_result(odp_ipsec_packet_result_t *result, odp_packet_t packet);
+ * @return 0      On success
+ * @retval <0     On failure
+ */
+int odp_ipsec_test_sa_update(odp_ipsec_sa_t sa, odp_ipsec_test_sa_param_t sa_param);
```
I think Petri's comment on this got somehow lost, so I am repeating it: The ID of the parameter that is modified could be a parameter of this function instead of contained in the sa_param parameter. Then odp_ipsec_sa_test_param_t could itself be the union of possible values.

asasidharan replied:
Agreed.

JannePeltonen commented on include/odp/api/spec/ipsec.h:
```diff
@@ -1715,6 +1786,32 @@ int odp_ipsec_result(odp_ipsec_packet_result_t *result, odp_packet_t packet);
+ *
+ * SA state update through this function may not be supported by all ODP
+ * implementations, ODP instances or SA instances or at every moment. This
+ * function may fail for unspecified reasons even when the capability
```
After dicsussing with Petri, it would be good to replace "may fail" with "may return failure". The former may perhaps incorrectly imply that the function might fail in some fatal, unrecoverable way (like segfault).

asasidharan replied:
Ack.

asasidharan commented on include/odp/api/spec/ipsec.h:
```diff
@@ -1715,6 +1786,32 @@ int odp_ipsec_result(odp_ipsec_packet_result_t *result, odp_packet_t packet);
+ * @return 0      On success
+ * @retval <0     On failure
+ */
+int odp_ipsec_test_sa_update(odp_ipsec_sa_t sa, odp_ipsec_test_sa_param_t sa_param);
```
Agreed. With V6, updated the API to have three parameters:
```
+int odp_ipsec_test_sa_update(odp_ipsec_sa_t sa,                                 
+                            odp_ipsec_test_sa_operation_t op,                   
+                            odp_ipsec_test_sa_param_t param); 
```

JannePeltonen commented on include/odp/api/spec/ipsec.h:
```diff
@@ -1715,6 +1782,35 @@ int odp_ipsec_result(odp_ipsec_packet_result_t *result, odp_packet_t packet);
+ */
+int odp_ipsec_test_sa_update(odp_ipsec_sa_t sa,
+			     odp_ipsec_test_sa_operation_t op,
+			     odp_ipsec_test_sa_param_t param);
```
I wonder if the param struct should be passed by reference rather than by value?

asasidharan replied:
Agreed. I will update this in v7.

JannePeltonen commented on include/odp/api/spec/ipsec.h:
```diff
@@ -192,6 +232,28 @@ typedef struct odp_ipsec_outbound_config_t {
+		 */
+		odp_bool_t antireplay_window_top;
+
+	} sa_update_params;
```
I think it would be better to name this sa_operations to match the name of the enum that got renamed odp_ipsec_test_sa_operation_t.

asasidharan replied:
Agreed. Renamed to sa_operations in v7.

JannePeltonen commented on test/validation/api/ipsec/ipsec.c:
```diff
@@ -351,6 +351,18 @@ int ipsec_check_esp_aes_gcm_256(void)
 				ODP_AUTH_ALG_AES_GCM, 0);
 }
 
+int ipsec_check_test_sa_update_seq_num(void)
```
This is a very minor thing, but it would be good to not place this function in the middle of the algorithm capability check functions but before or after them. It would keep the file slightly more easy to read as a whole.

asasidharan replied:
Agreed. Moved the function after the algorithm check capability functions.

JannePeltonen commented on test/validation/api/ipsec/ipsec.c:
```diff
@@ -866,6 +878,38 @@ static int ipsec_send_out_one(const ipsec_test_part *part,
+static void ipsec_pkt_seq_num_check(odp_packet_t pkt, uint32_t seq_num)
+{
+	uint32_t l3_off = odp_packet_l3_offset(pkt);
+	uint32_t l4_off = odp_packet_l4_offset(pkt);
```
This is incorrect. The value of l4 offset is undefined after outbound IPsec operation.

It would also be good to check whether the odp_packet_l3_offset() call succeeds.

asasidharan replied:
Agreed. Updated the function to compute l4_off as l3_off + IPv4 Header length.

Also added check for  return value of odp_packet_l3_offset().

JannePeltonen commented on test/validation/api/ipsec/ipsec.c:
```diff
@@ -866,6 +878,38 @@ static int ipsec_send_out_one(const ipsec_test_part *part,
+	uint32_t l4_off = odp_packet_l4_offset(pkt);
+	odph_ipv4hdr_t ip;
+
+	odp_packet_copy_to_mem(pkt, l3_off, sizeof(ip), &ip);
```
It would be good to check that the packet is an IPv4 packet. It would also be good to check the return value of the copy function.

asasidharan replied:
Ack.

JannePeltonen commented on test/validation/api/ipsec/ipsec.c:
```diff
@@ -866,6 +878,38 @@ static int ipsec_send_out_one(const ipsec_test_part *part,
+
+		odp_packet_copy_to_mem(pkt, l4_off, sizeof(esp), &esp);
+		CU_ASSERT_EQUAL(odp_be_to_cpu_32(esp.seq_no), seq_num);
+	} else {
```
It would be good to check that the packet indeed is AH at this point (and not e.g. UDP).

asasidharan replied:
Ack.

JannePeltonen commented on test/validation/api/ipsec/ipsec.c:
```diff
@@ -734,6 +746,44 @@ static int ipsec_send_out_one(const ipsec_test_part *part,
+	CU_ASSERT_EQUAL_FATAL(0, odp_packet_copy_to_mem(pkt, l3_off, sizeof(ip),
+							&ip));
+
+	if (ODPH_IPV4HDR_VER(ip.ver_ihl) == ODPH_IPV4) {
```
Now this function silently succeeds without checking the sequence number if IP version is not 4 or if IP protocol is not ESP or AH (e.g. IPsec with UDP encapsulation). In those case the function should fail (CU_FAIL() or similar).

asasidharan replied:
Agreed. Fixed by invoking CU_FAIL() for unexpected IP version and IP proto values in v8.

JannePeltonen commented:
Sorry, one more change: the param parameter of odp_ipsec_test_sa_update() should be a const pointer:

```
int odp_ipsec_test_sa_update(odp_ipsec_sa_t sa,
                             odp_ipsec_test_sa_operation_t op,
                             const odp_ipsec_test_sa_param_t *param);
```

With that change,
Reviewed-by: Janne Peltonen <janne.peltonen@nokia.com>

asasidharan commented:
> 
> 
> Sorry, one more change: the param parameter of odp_ipsec_test_sa_update() should be a const pointer:
> 
> ```
> int odp_ipsec_test_sa_update(odp_ipsec_sa_t sa,
>                              odp_ipsec_test_sa_operation_t op,
>                              const odp_ipsec_test_sa_param_t *param);
> ```

Agreed. Updated v9 by making param a const pointer.

Also thank you for your valuable suggestions and comments which helped in shaping this PR. I have added Reviewed-by tag to the commits.

psavol commented on include/odp/api/spec/ipsec.h:
```diff
@@ -71,6 +71,46 @@ typedef enum odp_ipsec_op_mode_t {
+	/** Update next sequence number
+	 *
+	 * The seq_num parameter is an outbound SA specific parameter.
+	 * Invoking the odp_ipsec_sa_test_update() API to update this
```
typo: odp_ipsec_sa_test_update() -> ..._test_sa_...

asasidharan replied:
Ack. Fixed with v10.

psavol commented on include/odp/api/spec/ipsec.h:
```diff
@@ -71,6 +71,46 @@ typedef enum odp_ipsec_op_mode_t {
+	/** Update highest authenticated sequence number
+	 *
+	 * The antireplay_window_top parameter is inbound SA specific.
+	 * Invoking the odp_ipsec_sa_test_update() API to update this
```
typo: odp_ipsec_sa_test_update() -> ..._test_sa_...

asasidharan replied:
Ack. Fixed with v10.

psavol commented:
API spec otherwise OK, but two typos in functions names. Add my reviewed-by to API patch for v10.

psavol commented:
Ping. This otherwise ready, but needs typo corrections and my reviewed-by to API patch.

asasidharan commented:
> 
> 
> Ping. This otherwise ready, but needs typo corrections and my reviewed-by to API patch.

@psavol  Thank you for pinging. I had missed to notice your previous comments until now.

With v10 fixed the typos and added  your reviewd-by tag to the API patch.
